cpp/src/arrow/gpu/cuda_context.cc

```diff
@@ -384,7 +384,7 @@ Result<std::shared_ptr<Buffer>> CudaMemoryManager::ViewBufferTo(
   if (to->is_cpu()) {
     // Device-on-CPU view
     ARROW_ASSIGN_OR_RAISE(auto address, GetHostAddress(buf->address()));
-    return std::make_shared<Buffer>(address, buf->size(), to, buf);
+    return std::make_shared<Buffer>(address, buf->size(), to, buf, DeviceType::CUDA_HOST);
   }
```

Comment:
I know this is existing code, but I don't think this is safe? It looks like we're taking the address to device memory and then just creating a new buffer wrapping that same address with a `CUDA_HOST` device?

This would only work if the device of `buf` was `CUDA_HOST` or `CUDA_MANAGED` which doesn't look to be the case here.

Reply:
Missed the `GetHostAddress` function which calls into `cuPointerGetAttribute` under the hood with `CU_POINTER_ATTRIBUTE_HOST_POINTER` to get a host accessible pointer. This will normally error, but that's fine since it works as intended. Resolving this.